[report] add audit log capture to help diagnose SELinux denials

a-palchikov opened the pull request:
This PR adds support for capturing SElinux denials from the host's audit log enriched with details specific to the installer.
The rules are only added for the duration (or part) of the operation to avoid introducing debugging state to the host. The rule captures only subset of syscalls responsible for creating/removing the filesystem resources as well as basic network operations.

This will hopefully aid in troubleshooting failed installs/joins on SELinux-enabled hosts.

Requires https://github.com/gravitational/planet/pull/589.

a-palchikov committed:
Add auditlog rules for installer for better visibility into SELinux
failures.

a-palchikov committed:
Move auditlog to a dedicated package

a-palchikov committed:
Only audit specific syscalls

a-palchikov committed:
Capture audit logs for each domain individually as the --subject flag
does not seem to be accumulative.

r0mant commented on lib/report/system.go:
```diff
@@ -158,3 +161,27 @@ func fetchEtc(name string) CollectorFunc {
+	var subjects []string
+	for _, domain := range auditlog.Domains {
+		subjects = append(subjects,
+			fmt.Sprint("/sbin/ausearch --success no --message all --raw --start yesterday --end today ", "--subject ", domain, " ;"),
```
Just curious how this time filter works. Will it collect the entire "yesterday"? Otherwise, if "today" has just started, there may not be enough data to work with.

a-palchikov replied:
I'll replace `today` with `now` although both seem to mean the same thing. Unless this is not what you meant.

r0mant commented on lib/system/auditlog/auditlog.go:
```diff
@@ -0,0 +1,105 @@
+/*
+Copyright 2018 Gravitational, Inc.
```
2020

r0mant commented on lib/system/auditlog/auditlog.go:
```diff
@@ -0,0 +1,105 @@
+	}
+}
+
+// AddDefaultRules adds default audit rules for the all known domains
```
Nit: just "for all known domains" I think.

r0mant commented on lib/system/auditlog/auditlog.go:
```diff
@@ -0,0 +1,105 @@
+	return nil
+}
+
+// RemoveRules removes audit rules previously with this package
```
... previously _configured_ (or _set up_) with this package?

r0mant commented on tool/gravity/cli/install.go:
```diff
@@ -725,6 +726,9 @@ func join(env *localenv.LocalEnvironment, environ LocalEnvironmentFactory, confi
 func TerminationHandler(interrupt *signals.InterruptHandler, printer utils.Printer) {
+	audit := auditlog.New()
+	audit.AddDefaultRules()   //nolint:errcheck
+	defer audit.RemoveRules() //nolint:errcheck
```
This can potentially not execute if the installer terminates in a some kind of ungraceful way. So I'm curious, what is the downside (if any) of these rules not being removed if that happens?

a-palchikov replied:
I haven't done exhaustive testing but batching and grouping fields and syscalls seems to be the recommended way to reduce kernel churn and limit the results.

Not sure what we can do else to counter the ungraceful termination other than explicitly considering or ignoring the rest of the signals that have termination as the default behavior and can be handled.

knisbet commented on lib/report/system.go:
```diff
@@ -158,3 +161,27 @@ func fetchEtc(name string) CollectorFunc {
+		)
+	}
+	script := fmt.Sprintf(`
+#!/bin/bash
```
Is bash guarenteed to be installed on all the distros we support? I imagine it is, but might be safer to make this work with /bin/sh?

knisbet commented on tool/gravity/cli/report.go:
```diff
@@ -19,23 +19,31 @@ package cli
 	}
-	err := report.Collect(context.TODO(), config, w)
-	return trace.Wrap(err)
+	return report.Collect(context.TODO(), config, w)
```
is there a reason this is no longer being wrapped?

a-palchikov committed:
Address review comments

a-palchikov committed:
Clean up audit rules on uninstall

a-palchikov committed:
Default to 'off' for SELinux until we've tested the user applications
enough to be sure.
Add override from environment variable for the '--selinux' flag.

a-palchikov committed:
Update planet to fix a bug on SELinux-enabled nodes.
Only export fio with SELinux label if SELinux has been enabled.

a-palchikov committed:
Update planet to release
